### ISSUE TITLE
an error about contiguous op

### ISSUE THREAD

**Issue by qianqing13579** [ROCmSoftwarePlatform/AMDMIGraphX#894]
when I use mobilenet_v2 to do inference,and set batch size to 16,I find that the result of each image is different! I use FP16 in release type.If I remove eliminate_contiguous from GPU target or insert a contiguous op between broadcast and gpu::add, the result will be right.

the model link: https://pan.baidu.com/s/1Jz5OO2pvKNKCIArx3G9fYA
password：htna

Here is my test code:

#include
#include
#include <migraphx/onnx.hpp>
#include <migraphx/gpu/target.hpp>
#include <migraphx/gpu/hip.hpp>
#include <migraphx/generate.hpp>
#include <migraphx/quantization.hpp>
#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>

using namespace cv;
using namespace cv::dnn;
using namespace std;
using namespace migraphx;

int main(int argc, char argv[])
{
// load
migraphx::program net= migraphx::parse_onnx("./mobilenet_v2.onnx");
std::pair<std::string, shape> inputAttribute=*(net.get_parameter_shapes().begin());
string inputName=inputAttribute.first;
shape inputShape=inputAttribute.second;
int N=inputShape.lens()[0];
int C=inputShape.lens()[1];
int H=inputShape.lens()[2];
int W=inputShape.lens()[3];
printf("input name:%s\n",inputName.c_str());
printf("input shape:%d,%d,%d,%d\n",N,C,H,W);

// compile
migraphx::quantize_fp16(net);
net.compile(migraphx::gpu::target{});

// input
int batchSize=N;
migraphx::parameter_map m;
for (std::pair<std::string, shape> x : net.get_parameter_shapes())
{
    m[x.first] = migraphx::gpu::to_gpu(migraphx::generate_argument(x.second));
}
vector<Mat> srcImages;
for(int i=0;i<batchSize;++i)
{
    Mat srcImage=imread("../Test.jpg");
    srcImages.push_back(srcImage);
}
Mat inputBlob;
blobFromImages(srcImages,inputBlob,0.0078125,cv::Size(W,H),cv::Scalar(127.5,127.5,127.5),true,false);
m[inputName]= migraphx::gpu::to_gpu(migraphx::argument{inputShape, (float*)inputBlob.data}); 

std::vector<argument> results = net.eval(m);
argument result  = migraphx::gpu::from_gpu(results[0]);
std::vector<float> logits;
result.visit([&](auto output) { logits.assign(output.begin(), output.end()); });

// output
int numberOfClasses=logits.size()/N;
for(int i=0;i<N;++i)
{
    printf("==========%d image output=============\n",i);
    int startIndex=numberOfClasses*i;

    std::vector<float> logit;
    for(int j=0;j<numberOfClasses;++j)
    {
        printf("%f\n",logits[startIndex+j]);
    }
    
}

return 0;
}

**Comment by qianqing13579** [comment]
I insert a contiguous op between broadcast and gpu::add by changing the code of  src/eliminate_contiguous.cpp.The changed file:

#include <migraphx/eliminate_contiguous.hpp>
#include <migraphx/program.hpp>
#include <migraphx/instruction.hpp>
#include <migraphx/iterator_for.hpp>
#include <migraphx/ranges.hpp>
#include <migraphx/stringutils.hpp>
#include <migraphx/op/contiguous.hpp>
#include <migraphx/op/identity.hpp>
#include <utility>

namespace migraphx {
inline namespace MIGRAPHX_INLINE_NS {

static bool try_compute_shape(instruction_ref ins, const std::vector<shape>& inputs)
{
    try
    {
        shape new_shape = ins->get_operator().compute_shape(inputs);
        // If the output shape is a standard shape, no need to try its output
        if(new_shape.standard())
        {
            return true;
        }

        // if no changes for the shape, the contiguous can also be removed
        if(new_shape == ins->get_shape())
        {
            return true;
        }

        auto outputs = ins->outputs();
        // If the current instruction has no output, it means it is the last
        // instruction and generates a non-standard output shape, and the last
        // output shape is different from the case with the contiguous operator
        if(outputs.empty())
        {
            return false;
        }

        for(auto output : outputs)
        {
            auto args = output->inputs();
            std::vector<shape> input_shapes(args.size());
            std::transform(args.begin(), args.end(), input_shapes.begin(), [&](auto& arg) {
                return (arg == ins) ? new_shape : arg->get_shape();
            });

            if(!try_compute_shape(output, input_shapes))
            {
                return false;
            }
        }
    }
    catch(...)
    {
        return false;
    }

    return true;
}

static bool try_compute_shape(instruction_ref ins, const std::vector<instruction_ref>& args)
{
    auto inputs = to_shapes(args);
    return try_compute_shape(ins, inputs);
}

void eliminate_contiguous::apply(module& p) const
{
    for(auto ins : iterator_for(p))
    {
        // return instruction should have inputs with standard shape
        if(ins->name() == "@return")
            continue;

        // Make a copy so we can modify it while we iterate
        auto args = ins->inputs();
        for(auto arg : ins->inputs())
        {
            if(arg->name() == op_name)
            {
                auto new_args = args;
                auto prev     = arg->inputs().front();

                ////////// insert contiguous between broadcastand gpu::add //////////
                if(prev->name()=="broadcast"&&ins->name()=="gpu::add")
                {
                    // printf("broadcast&&gpu::add\n");
                    continue;
                }
                /////////////////////////////////////////////////////////////////////


                replace(new_args, arg, prev);
                if(try_compute_shape(ins, new_args))
                {
                    instruction::replace_argument(ins, arg, prev);
                }
                else if(prev->can_eval())
                {
                    auto c = op::contiguous{};
                    auto r = c.compute(c.compute_shape({prev->get_shape()}), {prev->eval()});

                    auto l = p.add_literal(r.get_shape(), r.data());
                    p.replace_instruction(arg, l);
                }
            }
        }
    }
}

} // namespace MIGRAPHX_INLINE_NS
} // namespace migraphx

**Comment by qianqing13579** [comment]
I insert a contiguous op between broadcast and gpu::add by changing the code of  src/eliminate_contiguous.cpp.The changed file:

#include <migraphx/eliminate_contiguous.hpp>
#include <migraphx/program.hpp>
#include <migraphx/instruction.hpp>
#include <migraphx/iterator_for.hpp>
#include <migraphx/ranges.hpp>
#include <migraphx/stringutils.hpp>
#include <migraphx/op/contiguous.hpp>
#include <migraphx/op/identity.hpp>
#include <utility>

namespace migraphx {
inline namespace MIGRAPHX_INLINE_NS {

static bool try_compute_shape(instruction_ref ins, const std::vector<shape>& inputs)
{
    try
    {
        shape new_shape = ins->get_operator().compute_shape(inputs);
        // If the output shape is a standard shape, no need to try its output
        if(new_shape.standard())
        {
            return true;
        }

        // if no changes for the shape, the contiguous can also be removed
        if(new_shape == ins->get_shape())
        {
            return true;
        }

        auto outputs = ins->outputs();
        // If the current instruction has no output, it means it is the last
        // instruction and generates a non-standard output shape, and the last
        // output shape is different from the case with the contiguous operator
        if(outputs.empty())
        {
            return false;
        }

        for(auto output : outputs)
        {
            auto args = output->inputs();
            std::vector<shape> input_shapes(args.size());
            std::transform(args.begin(), args.end(), input_shapes.begin(), [&](auto& arg) {
                return (arg == ins) ? new_shape : arg->get_shape();
            });

            if(!try_compute_shape(output, input_shapes))
            {
                return false;
            }
        }
    }
    catch(...)
    {
        return false;
    }

    return true;
}

static bool try_compute_shape(instruction_ref ins, const std::vector<instruction_ref>& args)
{
    auto inputs = to_shapes(args);
    return try_compute_shape(ins, inputs);
}

void eliminate_contiguous::apply(module& p) const
{
    for(auto ins : iterator_for(p))
    {
        // return instruction should have inputs with standard shape
        if(ins->name() == "@return")
            continue;

        // Make a copy so we can modify it while we iterate
        auto args = ins->inputs();
        for(auto arg : ins->inputs())
        {
            if(arg->name() == op_name)
            {
                auto new_args = args;
                auto prev     = arg->inputs().front();

                ////////// insert contiguous between broadcastand gpu::add //////////
                if(prev->name()=="broadcast"&&ins->name()=="gpu::add")
                {
                    // printf("broadcast&&gpu::add\n");
                    continue;
                }
                /////////////////////////////////////////////////////////////////////


                replace(new_args, arg, prev);
                if(try_compute_shape(ins, new_args))
                {
                    instruction::replace_argument(ins, arg, prev);
                }
                else if(prev->can_eval())
                {
                    auto c = op::contiguous{};
                    auto r = c.compute(c.compute_shape({prev->get_shape()}), {prev->eval()});

                    auto l = p.add_literal(r.get_shape(), r.data());
                    p.replace_instruction(arg, l);
                }
            }
        }
    }
}

} // namespace MIGRAPHX_INLINE_NS
} // namespace migraphx

**Comment by scxiao** [comment]
I was able to run your program and can get the same outputs for different batches. Please check the attached file. 
[log.txt](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/files/6887937/log.txt)

A few notes:
1) I am using the latest develop branch. 
2) The log file attached printed out the compiled program, NO contiguous required between broadcast and add. 
3) For your test program in the first post, is that the program you are running? I see a syntax error in the following line:

```
std::pair<std::string, shape> inputAttribute=(net.get_parameter_shapes().begin());
```
I think it should be:

```
std::pair<std::string, shape> inputAttribute= *net.get_parameter_shapes().begin();
 
```
Not sure whether it is related.

**Comment by qianqing13579** [comment]
@scxiao ,can you please test the efficientnet model? I find the same problem for different batches in my machine. strange!
the model link: https://pan.baidu.com/s/1DSOz8lgrB_7mP0p5qGHTiQ 
password：cz3k

**Comment by scxiao** [comment]
It works fine in my computer. What is your output?

**Comment by qianqing13579** [comment]
@scxiao ,here is my project which includes my code ,test image and output,and I can get different outputs for different batches.
the project link: https://pan.baidu.com/s/1LC1lISlUmR8SwbL5MRMI_w
password: k4lg

**Comment by qianqing13579** [comment]
@scxiao , I have found out the bug. I tried to set the bias to 0 of the nary_broadcast_impl function in nary.hpp,and the result will be right. so the nary_broadcast_impl function has problems in computing bias.
![image](https://user-images.githubusercontent.com/8730200/128481854-8d5ab5d5-27ea-403a-b168-215c395e6c4c.png)

**Comment by qianqing13579** [comment]
@scxiao ,here is a small model  that show the problem more easily. The file includes code,image and output of each instrcution. The result of adding contiguous is in output_addcontiguous.log. The output of the following instruction is different by comparing with outpu.log.
main:@17 = gpu::add_clip(main:@9,main:@13,main:@15,main:@16,main:@14) -> half_type, {2, 32, 4, 4}, {512, 16, 4, 1}

the project link: https://pan.baidu.com/s/1LSQtapa7aB3sp7jn50v__w 
password: k9i1

**Comment by scxiao** [comment]
@qianqing13579 Thanks a lot for the information and we are able to reproduce the issue and got a PR #906 for that. I am wondering whether you can try whether the PR resolved your issue? Thanks

**Comment by qianqing13579** [comment]
@scxiao  I have tried,but it does not work.

**Comment by cagery** [comment]
Have you tried without quantizing the model to fp16? @qianqing13579

**Comment by qianqing13579** [comment]
@cagery ,I have tried,but it does not work.I think the problem #894 is different from #906.

**Comment by qianqing13579** [comment]
@scxiao ,@cagery, I open a new issue #908. I will close this issue

**Comment by qianqing13579** [comment]
@scxiao ,@cagery, I open a new issue #908. I will close this issue

**Comment by qianqing13579** [comment]
@scxiao ,@cagery, I open a new issue #908. I will close this issue
